In [2]:
import psycopg2

In [6]:
hostname = 'localhost'
database = 'postgres'
username = 'postgres'
pwd = '***'
port_id = 5432

In [7]:
conn = psycopg2.connect(
            host = hostname,
            dbname = database,
            user = username,
            password = pwd,
            port = port_id)

In [8]:
conn

<connection object at 0x000001CE75740040; dsn: 'user=postgres password=xxx dbname=postgres host=localhost port=5432', closed: 0>

In [11]:
# DB-API: The python standard for connecting to Relational DBs

# Creating a cursor object to query in python
cur = conn.cursor()
cur

<cursor object at 0x000001CE7800CD60; closed: 0>

In [12]:
type(cur)

psycopg2.extensions.cursor

<b>This is how we do our queries. For each time when we want to send a query, we will send it via the cursor.</b>

<b>EXAMPLE</b>

In [32]:
cur.execute("""SELECT *
                FROM finance
                LIMIT 10""")

In [33]:
for i in cur.fetchall():
    product_id, listing_price, sale_price, discount, revenue = i
    print(f"[{product_id}]: Listing Price: {listing_price}, Sale Price: {sale_price}, Discount: {discount}, Revenue: {revenue}")

[AH2430]: Listing Price: None, Sale Price: None, Discount: None, Revenue: None
[G27341]: Listing Price: 75.99, Sale Price: 37.99, Discount: 0.5, Revenue: 1641.17
[CM0081]: Listing Price: 9.99, Sale Price: 5.99, Discount: 0.4, Revenue: 398.93
[B44832]: Listing Price: 69.99, Sale Price: 34.99, Discount: 0.5, Revenue: 2204.37
[D98205]: Listing Price: 79.99, Sale Price: 39.99, Discount: 0.5, Revenue: 5182.7
[B75586]: Listing Price: 47.99, Sale Price: 19.2, Discount: 0.6, Revenue: 1555.2
[CG4051]: Listing Price: 47.99, Sale Price: 23.99, Discount: 0.5, Revenue: 86.36
[CM0080]: Listing Price: 9.99, Sale Price: 5.99, Discount: 0.4, Revenue: 75.47
[B75990]: Listing Price: 55.99, Sale Price: 27.99, Discount: 0.5, Revenue: 806.11
[EE5761]: Listing Price: 65.99, Sale Price: 39.59, Discount: 0.4, Revenue: 2779.22


In [41]:
import pandas as pd

#### INFORMATION TABLE - Product Name, Product ID, and Description

In [72]:
df1 = pd.read_csv(r"C:\Users\Prateek\OneDrive\Desktop\datasets\info_v2.csv")
df1.head()

product_name product_id  \
0                                               NaN     AH2430   
1              Women's adidas Originals Sleek Shoes     G27341   
2                 Women's adidas Swim Puka Slippers     CM0081   
3  Women's adidas Sport Inspired Questar Ride Shoes     B44832   
4          Women's adidas Originals Taekwondo Shoes     D98205   

                                         description  
0                                                NaN  
1  A modern take on adidas sport heritage, tailor...  
2  These adidas Puka slippers for women's come wi...  
3  Inspired by modern tech runners, these women's...  
4  This design is inspired by vintage Taekwondo s...

#### FINANCE TABLE - Product ID, Listing Price, Sale Price, Discount, and Revenue

In [43]:
df2 = pd.read_csv(r"C:\Users\Prateek\OneDrive\Desktop\datasets\finance.csv")
df2.head()

product_id  listing_price  sale_price  discount  revenue
0     AH2430            NaN         NaN       NaN      NaN
1     G27341          75.99       37.99       0.5  1641.17
2     CM0081           9.99        5.99       0.4   398.93
3     B44832          69.99       34.99       0.5  2204.37
4     D98205          79.99       39.99       0.5  5182.70

#### REVIEWS - Product ID, Rating, and Reviews

In [44]:
df3 = pd.read_csv(r"C:\Users\Prateek\OneDrive\Desktop\datasets\reviews_v2.csv")
df3.head()

product_id  rating  reviews
0     AH2430     NaN      NaN
1     G27341     3.3     24.0
2     CM0081     2.6     37.0
3     B44832     4.1     35.0
4     D98205     3.5     72.0

#### TRAFFIC TABLE - Product ID, Last Visited Timestamp

In [45]:
df4 = pd.read_csv(r"C:\Users\Prateek\OneDrive\Desktop\datasets\traffic_v3.csv")
df4.head()

product_id         last_visited
0     AH2430  2018-05-19 15:13:00
1     G27341  2018-11-29 16:16:00
2     CM0081  2018-02-01 10:27:00
3     B44832  2018-09-07 20:06:00
4     D98205  2019-07-18 15:26:00

#### BRANDS TABLE - Product ID, Brand

In [46]:
df5 = pd.read_csv(r"C:\Users\Prateek\OneDrive\Desktop\datasets\brands_v2.csv")
df5.head()

product_id   brand
0     AH2430     NaN
1     G27341  Adidas
2     CM0081  Adidas
3     B44832  Adidas
4     D98205  Adidas

In [48]:
#!pip3 install ipython-sql

In [49]:
#%load_ext sql

<h3>COUNTING NON-MISSING VALUES</h3>

+ Joining the following tables using inner join on the 'product_id' column:

    > information
\
    > finance
\
    > traffic

+ Counting the number of rows in description, listing price, and last visited columns

In [40]:
# Connecting to the database
%%sql
postgresql://postgres:***@localhost/postgres

SELECT COUNT(*) AS total_rows, 
    COUNT(i.description) AS count_description, 
    COUNT(f.listing_price) AS count_listing_price, 
    COUNT(t.last_visited) AS count_last_visited 
FROM information AS i
INNER JOIN finance AS f
    ON i.product_id = f.product_id
INNER JOIN traffic AS t
    ON t.product_id = f.product_id;

1 rows affected.


[(3179, 3117, 3120, 2928)]

###### The database contains 3,179 products in total, and the column 'last_visited' is missing more than five percent of its values.

<h3>NIKE Vs. ADIDAS Pricing</h3>

+ Exploring where the price points of Nike and Adidas differ to gauge the company's stock range and customer market.

In [52]:
%%sql

/* Running a query to produce a distribution of the 'listing_price' and the count for each price, grouped by brand */

SELECT b.brand, f.listing_price::integer, COUNT(f.*)
FROM finance AS f
INNER JOIN brands AS b 
    ON f.product_id = b.product_id
WHERE listing_price > 0
GROUP BY b.brand, f.listing_price
ORDER BY listing_price DESC;

 * postgresql://postgres:***@localhost/postgres
77 rows affected.


[('Adidas', 300, 2),
 ('Adidas', 280, 4),
 ('Adidas', 240, 5),
 ('Adidas', 230, 8),
 ('Adidas', 220, 11),
 ('Nike', 200, 1),
 ('Adidas', 200, 8),
 ('Nike', 190, 2),
 ('Adidas', 190, 7),
 ('Nike', 180, 4),
 ('Adidas', 180, 34),
 ('Nike', 170, 14),
 ('Adidas', 170, 27),
 ('Nike', 160, 31),
 ('Adidas', 160, 28),
 ('Adidas', 150, 41),
 ('Nike', 150, 6),
 ('Adidas', 140, 36),
 ('Nike', 140, 12),
 ('Adidas', 130, 96),
 ('Nike', 130, 12),
 ('Adidas', 120, 115),
 ('Nike', 120, 16),
 ('Nike', 110, 17),
 ('Adidas', 110, 91),
 ('Nike', 100, 14),
 ('Adidas', 100, 72),
 ('Adidas', 96, 2),
 ('Nike', 95, 1),
 ('Nike', 90, 13),
 ('Adidas', 90, 89),
 ('Adidas', 86, 7),
 ('Nike', 85, 5),
 ('Adidas', 85, 1),
 ('Adidas', 80, 322),
 ('Nike', 80, 16),
 ('Nike', 79, 1),
 ('Adidas', 76, 149),
 ('Adidas', 75, 1),
 ('Nike', 75, 7),
 ('Adidas', 70, 87),
 ('Nike', 70, 4),
 ('Adidas', 66, 102),
 ('Nike', 65, 1),
 ('Adidas', 63, 1),
 ('Nike', 60, 2),
 ('Adidas', 60, 211),
 ('Adidas', 56, 174),
 ('Adidas', 55, 2),
 ('Adidas', 53, 43),
 ('Nike', 50, 5),
 ('Adidas', 50, 183),
 ('Adidas', 48, 42),
 ('Nike', 48, 1),
 ('Adidas', 46, 163),
 ('Nike', 45, 3),
 ('Adidas', 45, 1),
 ('Adidas', 43, 51),
 ('Adidas', 40, 81),
 ('Nike', 40, 1),
 ('Adidas', 38, 24),
 ('Adidas', 36, 25),
 ('Adidas', 33, 24),
 ('Adidas', 30, 37),
 ('Nike', 30, 2),
 ('Adidas', 28, 38),
 ('Adidas', 27, 18),
 ('Adidas', 25, 28),
 ('Adidas', 23, 1),
 ('Adidas', 20, 8),
 ('Adidas', 18, 4),
 ('Adidas', 16, 4),
 ('Adidas', 15, 27),
 ('Adidas', 13, 27),
 ('Adidas', 12, 1),
 ('Adidas', 10, 11),
 ('Adidas', 9, 1)]

##### There are 77 unique prices for the products in the database.

<h3>LABELING PRICE CHANGES</h3>

+ Assigning labels to different price ranges while grouping by 'brand' and 'label'. Also, including the total revenue for each price range and brand. 

In [63]:
%%sql

/* Calculating the median of the listing price */

SELECT PERCENTILE_CONT(0.50) WITHIN GROUP (ORDER BY listing_price) AS median
  FROM finance;

 * postgresql://postgres:***@localhost/postgres
1 rows affected.


[(59.99,)]

In [65]:
%%sql

SELECT b.brand, COUNT(f.*), SUM(f.revenue) as total_revenue,
CASE WHEN f.listing_price < 46 THEN 'Budget'
    WHEN f.listing_price >= 46 AND f.listing_price < 74 THEN 'Average'
    WHEN f.listing_price >= 90 AND f.listing_price < 129 THEN 'Expensive'
    ELSE 'Elite' END AS price_category
FROM finance AS f
INNER JOIN brands AS b 
    ON f.product_id = b.product_id
WHERE b.brand IS NOT NULL
GROUP BY b.brand, price_category
ORDER BY total_revenue DESC;

 * postgresql://postgres:***@localhost/postgres
8 rows affected.


[('Adidas', 876, 5781879.049999994, 'Elite'),
 ('Adidas', 845, 2728323.9599999986, 'Average'),
 ('Adidas', 280, 1859417.85, 'Expensive'),
 ('Adidas', 574, 1156998.2199999988, 'Budget'),
 ('Nike', 360, 595583.8099999992, 'Budget'),
 ('Nike', 124, 151026.07000000004, 'Elite'),
 ('Nike', 48, 49292.670000000006, 'Expensive'),
 ('Nike', 13, 6380.709999999999, 'Average')]

We can see that Adidas items generate more revenue regardless of the price category. The <b>Elite</b> category generates the highest revenue, so, the company can potentially increase revenue by shifting their stock to have a larger proportion of these products.

+ <i>The median listing price at 25th percentile was at <b>~46</b>, which is benchmarked as <b>Budget</b></i>


+ <i>The median listing price at 50th percentile was at <b>~60</b>, which is benchmarked as <b>Average</b></i>


+ <i>The median listing price at 75th percentile was at <b>~90</b>, which is benchmarked as <b>Expensive</b></i> 


+ <i>Everything <b>>90</b> was benchmarked as <b>Elite</b></i> 

<h3>AVERAGE DISCOUNT BY BRAND</h3>

In [66]:
%%sql

SELECT b.brand, AVG(f.discount) * 100 AS average_discount
FROM brands AS b
INNER JOIN finance AS f 
    ON b.product_id = f.product_id
GROUP BY b.brand
HAVING b.brand IS NOT NULL
ORDER BY average_discount;

 * postgresql://postgres:***@localhost/postgres
2 rows affected.


[('Nike', 0.0), ('Adidas', 33.452427184465606)]

##### Strangely, no dicount is offered on Nike products. In comparison, not only do Adidas products generate the most revenue, but these products are also heavily discounted.

To improve revenue further, the company could try to reduce the amount of discount offered on Adidas products, and monitor sales volume to see if it remains stable. Alternatively, it could try offering a small discount on Nike products. This would reduce average revenue for these products, but may increase revenue overall if there is an increase in the volume of Nike products sold.

<h3>CORRELATION BETWEEN REVIEWS AND REVENUE</h3

In [68]:
%%sql

SELECT corr(r.reviews, f.revenue) AS review_revenue_corr
FROM reviews AS r
INNER JOIN finance AS f 
    ON r.product_id = f.product_id;

 * postgresql://postgres:***@localhost/postgres
1 rows affected.


[(0.6518512283481301,)]

##### There is a positive correlation between revenue and reviews

<h3>FOOTWEAR PRODUCT PERFORMANCE</h3>

+ Creating a <b>Common Table Expression (CTE)</b> that filters description for keywords, then uses the results to find out how much of the company's stock consists of footwear products and the median revenue generated by these items.

In [70]:
%%sql

WITH footwear AS
(
    SELECT i.description, f.revenue
    FROM information AS i
    INNER JOIN finance AS f 
        ON i.product_id = f.product_id
    WHERE i.description ILIKE '%shoe%'
        OR i.description ILIKE '%trainer%'
        OR i.description ILIKE '%foot%'
        AND i.description IS NOT NULL
)

SELECT COUNT(*) AS num_footwear_products, 
    percentile_disc(0.5) WITHIN GROUP (ORDER BY revenue) AS median_footwear_revenue
FROM footwear;

 * postgresql://postgres:***@localhost/postgres
1 rows affected.


[(2700, 3118.36)]

##### There were 3,117 without missing values for description. Of those, 2,700 are footwear products, which accounts for around 85% of the company's stock. They also generate a median revenue of over 3100 dollars.

<h3>CLOTHING PRODUCT PERFORMANCE</h3>

+ Since there is no point of reference for the median revenue generated by footwear, we can compare that of the clothing products to see if footwear revenue is good or bad.

In [74]:
%%sql

WITH footwear AS
(
    SELECT i.description, f.revenue
    FROM information AS i
    INNER JOIN finance AS f 
        ON i.product_id = f.product_id
    WHERE i.description ILIKE '%shoe%'
        OR i.description ILIKE '%trainer%'
        OR i.description ILIKE '%foot%'
        AND i.description IS NOT NULL
)

SELECT COUNT(i.*) AS num_clothing_products, 
    percentile_disc(0.5) WITHIN GROUP (ORDER BY f.revenue) AS median_clothing_revenue
FROM information AS i
INNER JOIN finance AS f on i.product_id = f.product_id
WHERE i.description NOT IN (SELECT description FROM footwear);

 * postgresql://postgres:***@localhost/postgres
1 rows affected.


[(417, 503.82)]

##### Since the median revenue generated by clothing products is around 500 Dollars, as compared to 3100 Dollars from footwear, we can definitely say that footwear products should be prioritized and capiltalized better for higher revenue.